torch/_decomp/decompositions.py

```diff
@@ -3455,6 +3455,59 @@ def arange_start(
     )
 
 
+def _multi_margin_loss_shape_check(ndims, input, target):
+    valid_inputs = (
+        (ndims == 2 and input.size(1) != 0)
+        or (ndims == 1 and input.size(0) != 0)
+        or ndims == 0
+    )
+    if ndims <= 1:
+        nframe = 1
+        dim = 1 if ndims == 0 else input.size(0)
+    else:
+        nframe = input.size(0)
+        dim = input.size(1)
+    torch._check(
+        valid_inputs,
+        lambda: f"Expected non-empty vector or matrix with optional 0-dim batch size, but got: {input.shape}",
+    )
+    torch._check(
+        valid_inputs and target.ndim <= 1 and target.numel() == nframe,
+        lambda: f"inconsistent target size, got: {target.shape}",
+    )
+    return nframe, dim
+
+
+@register_decomposition(aten.multi_margin_loss)
+@aten.multi_margin_loss.default.py_impl(DispatchKey.Autograd)
+@out_wrapper()
+def multi_margin_loss(
+    input: Tensor,
+    target: Tensor,
+    p: NumberType = 1,
+    margin: NumberType = 1,
+    weight: Optional[Tensor] = None,
+    reduction: int = Reduction.MEAN.value,
+) -> Tensor:
+    ndims = input.ndim
+    torch._check(p == 1 or p == 2, lambda: "only p == 1 and p == 2 supported")
+    _multi_margin_loss_shape_check(ndims, input, target)
+    input = torch.atleast_2d(input)
+    target = torch.atleast_1d(target).unsqueeze(1)
+    if weight is not None:
+        weight = torch.atleast_1d(weight)
+    size = input.size(1)
+    mask = torch.arange(size, device=input.device) != target
+    u = torch.gather(input, dim=1, index=target)
+    z = margin - u + input
+    z = z.clamp_min(0)
+    z = z if p == 1 else z * z
+    if weight is not None:
+        z = z * weight[target]
+    loss = z.where(mask, 0).mean(dim=1)
+    return apply_loss_reduction(loss, reduction)
```

Comment:
This isn't a "split reduction" per se, it's just two reductions one after the other that inductor can't fuse. I can get inductor to generate a single kernel by writing this as:
```python
    masked = z.where(mask, 0)
    if reduction == Reduction.MEAN.value:
        return masked.mean()
    elif reduction == Reduction.SUM.value:
        return masked.sum() / masked.shape[1]
    else:
        return masked.mean(dim=1)
```

However, it's still slower than eager. Looking at nvprof though it looks like the kernel itself is running faster than eager it's just down to overhead.

Reply:
Ah, yeah, I tried that one. I sent a simplified version of the decomposition to Nikita internally. My point that, even with this trick, the benchmarks that Nikita was running with `shape = (500, 500)`, were creating a split reduction.

Reply:
I just benchmarked and `(500, 500)` without split reduction is more than 3x slower...